### PR TITLE
Modularizing KOS, my take

### PR DESCRIPTION
**Pull request by pcercuei**
A rewrite of https://github.com/KallistiOS/KallistiOS/pull/335.

The bulk of the changes are by GyroVorbis, so he should probably sign-off the commits as well (or give me the permission to add his sign-off).

Compiling the "hello" example with just `KOS_INIT_FLAGS(INIT_IRQ | INIT_THD_PREEMPT | INIT_MALLOCSTATS | INIT_FS_ROMDISK);`, in which case the maple code and VMU FS code are garbage-collected, the resulting (stripped) hello.elf drops from 156 KiB to 116 KiB.

### CODE REVIEW

**Review by gyrovorbis — APPROVED** [review]
Looks good. This is definitely the PR we want over my old one. Cleaned up a bunch of stuff, made `maple_wait_scan()` only get called when at least one maple device is enabled, changed `likely` and `unlikely` to match the Linux macros, etc.

File: include/kos/init.h [hunk 1]
```diff
@@ -87,18 +91,18 @@ extern void * __kos_romdisk;
...
+
+ #define INIT_NONE        0x00000000  /**< \brief Don't init optional things */
+ #define INIT_IRQ         0x00000001  /**< \brief Enable IRQs at startup */
+ /* Preemptive mode is the only mode now. Keeping define for compatability. */
+ #define INIT_THD_PREEMPT 0x00000002  /**< \deprecated Already default mode */
+ #define INIT_NET         0x00000004  /**< \brief Enable built-in networking */
+ #define INIT_MALLOCSTATS 0x00000008  /**< \brief Enable malloc statistics */
+ #define INIT_QUIET       0x00000010  /**< \brief Disable dbgio */
+ #define INIT_EXPORT      0x00000020  /**< \brief Export kernel symbols */
+ #define INIT_FS_ROMDISK  0x00000040  /**< \brief Enable support for romdisks */
```
Comment (ljsebald) [inline include/kos/init.h:105]:
```suggestion
#define INIT_DEFAULT    (INIT_IRQ | INIT_THD_PREEMPT | INIT_FS_ROMDISK | \
                         INIT_DEFAULT_ARCH)

#define INIT_NONE        0x00000000  /**< \brief Don't init optional things */
#define INIT_IRQ         0x00000001  /**< \brief Enable IRQs at startup */
/* Preemptive mode is the only mode now. Keeping define for compatability. */
#define INIT_THD_PREEMPT 0x00000002  /**< \deprecated Already default mode */
#define INIT_NET         0x00000004  /**< \brief Enable built-in networking */
#define INIT_MALLOCSTATS 0x00000008  /**< \brief Enable malloc statistics */
#define INIT_QUIET       0x00000010  /**< \brief Disable dbgio */
#define INIT_EXPORT      0x00000020  /**< \brief Export kernel symbols */
#define INIT_FS_ROMDISK  0x00000040  /**< \brief Enable support for romdisks */
```

**Review by ljsebald — CHANGES_REQUESTED** [review]
One tiny formatting fixup...

File: include/kos/init.h [hunk 2]
```diff
@@ -25,44 +31,42 @@
...
+     KOS_INIT_FLAG(flags, INIT_NET, arch_init_net); \
+     KOS_INIT_FLAG(flags, INIT_NET, net_shutdown); \
+     KOS_INIT_FLAG(flags, INIT_NET, bba_la_init); \
+     KOS_INIT_FLAG(flags, INIT_NET, bba_la_shutdown); \
+     KOS_INIT_FLAG(flags, INIT_FS_ROMDISK, fs_romdisk_init); \
+     KOS_INIT_FLAG(flags, INIT_FS_ROMDISK, fs_romdisk_shutdown); \
+     KOS_INIT_FLAG(flags, INIT_EXPORT, export_init); \
+     KOS_INIT_FLAGS_ARCH(flags)
 
 /** \brief  The init flags. Do not modify this directly! */
```
Comment (gyrovorbis) [inline include/kos/init.h:61]:
```suggestion
/** \cond */
```

File: include/kos/init.h [hunk 3]
```diff
@@ -25,44 +31,42 @@
...
 extern uint32 __kos_init_flags;
```
Comment (gyrovorbis) [inline include/kos/init.h:63]:
```suggestion
/** \endcond */

```

**Review by gyrovorbis — COMMENTED** [review]
Crap just saw a couple things I missed before.

**Comment by pcercuei** [comment]
Updated and force-pushed.

**Review by ljsebald — APPROVED** [review]

**Pull request merged**